packages/go_router/lib/src/builder.dart

```diff
@@ -68,6 +68,10 @@ class RouteBuilder {
   final Map<GlobalKey<NavigatorState>, HeroController> _goHeroCache =
       <GlobalKey<NavigatorState>, HeroController>{};
 
+  /// State for any active stateful shell routes
+  final Map<ShellRouteBase, Object?> _shellRouteState =
```

Comment:
Well, if we want to avoid "weird" ways of wiring up StatefulShellRouteState with StatefulNavigationShell, we kind of need to have this here as I see it. Either that or return to the previous way of using an InheritedWidget and a somewhat more complex builder/pageBuilder setup.

Reply:
No, what I meant is to have one state update system that have a general API to take necessary input during the build step. the system hold all these states and update accoudingly. I don't think anything would change here, the problem with the current structure is that if we update the build logic, we need to make sure all three state also need to be update correclty.

Reply:
Ah, sorry, misunderstood you. Yes, I was kind of thinking about this as well, but this seemed like a pretty large change, so I was thinking this was something that could be done in stages. But I think some kind of centralized state management would good.